gateway-api-integrator/lib/charms/gateway_api_integrator/v0/gateway_route.py

```diff
@@ -0,0 +1,702 @@
+    )
+
+    # 2.To initialize the requirer with no parameters, i.e
+    # self.gateway_route_requirer = GatewayRouteRequirer(self)
+    # This will simply initialize the requirer class and it won't perfom any action.
+
+    # Afterwards regardless of how you initialized the requirer you can call the
+    # provide_gateway_route_requirements method anywhere in your charm to update the requirer data.
+    # The method takes the same number of parameters as the requirer class.
+    # provide_gateway_route_requirements(name=, model=, ...)
+
+    self.framework.observe(
+        self.framework.on.config_changed, self._on_config_changed
+    )
+    self.framework.observe(
+        self.gateway_route_requirer.on.ready, self._on_endpoints_ready
+    )
+    self.framework.observe(
+        self.gateway_route_requirer.on.removed, self._on_endpoints_removed
+    )
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        self.gateway_route_requirer.provide_gateway_route_requirements(...)
+
+    def _on_endpoints_ready(self, _: EventBase) -> None:
+        # Handle endpoints ready event
+        ...
+
+    def _on_endpoints_removed(self, _: EventBase) -> None:
+        # Handle endpoints removed event
+        ...
+
+## Using the library as the provider
+The provider charm should expose the interface as shown below:
+```yaml
+provides:
+    gateway-route:
+        interface: gateway-route
+```
+Note that this interface supports relating to multiple endpoints.
+
+Then, to initialise the library:
+```python
+from charms.haproxy.v1.haproxy_route import GatewayRouteProvider
+
+class SomeCharm(CharmBase):
+    self.gateway_route_provider = GatewayRouteProvider(self)
+    self.framework.observe(
+        self.gateway_route_provider.on.data_available, self._on_gateway_route_data_available
+    )
+
+    def _on_gateway_route_data_available(self, event: EventBase) -> None:
+        data = self.gateway_route_provider.get_data(self.gateway_route_provider.relation)
+        ...
+"""
+
+import json
+import logging
+from functools import partial
+from typing import Annotated, Any, MutableMapping, Optional, cast
+
+from ops import CharmBase, ModelError, RelationBrokenEvent
+from ops.charm import CharmEvents
+from ops.framework import EventBase, EventSource, Object
+from ops.model import Relation
+from pydantic import (
+    AnyHttpUrl,
+    BaseModel,
+    BeforeValidator,
+    ConfigDict,
+    Field,
+    ValidationError,
+)
+from pydantic.dataclasses import dataclass
+
+# The unique Charmhub library identifier, never change it
+LIBID = "e9aa842e-9df3-4ae9-affc-2ed3dcf12788"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+GATEWAY_ROUTE_RELATION_NAME = "gateway-route"
+GATEWAY_CONFIG_INVALID_CHARACTERS = "\n\t#\\'\"\r$ "
```

Comment:
These are haproxy specific characters that are not relevant in gateway API language do we can remove this